src/main/java/slimeknights/tconstruct/smeltery/tileentity/CastingTileEntity.java

```diff
@@ -0,0 +1,232 @@
+      setInventorySlotContents(INPUT, stack);
+    }
+    // take item out
+    else {
+      // take out stack 1 if something is there, 0 otherwise
+      int slot = isStackInSlot(OUTPUT) ? OUTPUT : INPUT;
+
+      // Additional info: Only 1 item can be put into the casting block usually, however recipes
+      // can have ItemStacks with stacksize > 1 as output
+      // we therefore spill the whole contents on extraction.
+      ItemStack stack = getStackInSlot(slot);
+      if (slot == OUTPUT) {
+        // fire player smelt event?
+      }
+      ItemHandlerHelper.giveItemToPlayer(player, stack);
+      setInventorySlotContents(slot, ItemStack.EMPTY);
+
+      // send a block update for the comparator, needs to be done after the stack is removed
+      if (slot == OUTPUT) {
+        this.getWorld().notifyNeighborsOfStateChange(this.pos, this.getBlockState().getBlock());
+      }
+    }
+  }
+
+
+  @Override
+  @Nonnull
+  public int[] getSlotsForFace(Direction side) {
+    return new int[]{INPUT, OUTPUT};
+  }
+
+  @Override
+  public boolean canInsertItem(int index, ItemStack itemStackIn, @Nullable Direction direction) {
+    return index == INPUT && !isStackInSlot(OUTPUT);
+  }
+
+  @Override
+  public boolean canExtractItem(int index, ItemStack stack, Direction direction) {
+    return index == OUTPUT;
+  }
+
+  @Override
+  public void tick() {
+    // no recipe
+    if (recipe == null) {
+      return;
+    }
+    // fully filled
+    if (tank.getFluidAmount() == tank.getCapacity() && !tank.getFluid().isEmpty()) {
+      timer++;
+      if (!getWorld().isRemote) {
+        if (timer >= recipe.getCoolingTime()) {
+          if (recipe.isConsumed()) {
+            setInventorySlotContents(INPUT, ItemStack.EMPTY);
+            // Fire particles
+          }
+          if (recipe.switchSlots()) {
+            setInventorySlotContents(INPUT, recipe.getRecipeOutput());
+          }
+          else {
+            setInventorySlotContents(OUTPUT, recipe.getRecipeOutput());
+          }
+          getWorld().playSound(null, pos, SoundEvents.BLOCK_LAVA_EXTINGUISH, SoundCategory.AMBIENT, 0.07f, 4f);
+
+          reset();
+
+          getWorld().notifyNeighborsOfStateChange(this.pos, this.getBlockState().getBlock());
+        }
+      }
+      else if (getWorld().rand.nextFloat() > 0.9f) {
+        world.addParticle(ParticleTypes.SMOKE, pos.getX() + getWorld().rand.nextDouble(), pos.getY() + 1.1d, pos.getZ() + getWorld().rand.nextDouble(), 0.0D, 0.0D, 0.0D);
+      }
+    }
+  }
+
+  protected AbstractCastingRecipe findRecipe() {
+    if (this.recipe != null && this.recipe.matches(crafting, getWorld())) {
+      return this.recipe;
+    }
+    AbstractCastingRecipe castingRecipe = getWorld().getRecipeManager().getRecipe(this.recipeType, crafting, getWorld()).orElse(null);
+
+    return castingRecipe;
+  }
+
+  public int initNewCasting(Fluid fluid, IFluidHandler.FluidAction action) {
+    if (this.fluid == Fluids.EMPTY) {
```

Comment:
I would just set the fluid regardless. Its not really relevant besides in this method, and we don't want to risk a mod simulating with one fluid, then being unable to cast a second one. This is likely causing your bugs with the inconsistently failing recipes, as if you set the fluid and never find a recipe, you are stuck with that fluid forever.

Reply:
Nope. This isn't the cause of the sporadic failures... very strange.

Reply:
In any case, this still is wrong, so remove the condition.